### ISSUE TITLE
After updating to 1.8, the power menu cannot be called out when the screen is locked.

### ISSUE THREAD

**Issue by YamatoRyou** [KieronQuinn/ClassicPowerMenu#143]
OS: Pixel OS (Android 14)
Model: OnePlus 9RT (MT2110)

**The system has installed Magisk + LSPosed, the application has been granted root permissions, and the application's Xposed module has been activated.**
**The system is set to fingerprint + pattern to unlock the phone.**

After updating to 1.8, if the phone is locked but the screen is on and the power button is pressed and held, the power menu will only flash for a moment and then disappear. The problem will recur if the power button is pressed and held repeatedly.
The current known solution is to roll back to 1.7.2.

**Comment by JLJTGR** [comment]
I have a simple Magisk rooted Pixel 8 Pro (A14 - AP2A.240805.005)and have the same issue.

But on top of that, once the device is unlocked... it might still happen up to 4 times.  Sometimes it will show up properly, but I will randomly have to retry several times to get a power menu.

Here is a slightly cropped video of it working correctly one time, then needing three attempts to work afterwards:
<details>
  <summary>Spoiler: Click to reveal...</summary>

https://github.com/user-attachments/assets/3afb423f-0abe-455e-9c1a-2295f8e16d8b
</details>

**Comment by KieronQuinn** [comment]
It's opening 100% of the time for me on my Pixel 8 Pro. Can you get a log, see if it's crashing the times it's not opening?

**Comment by JLJTGR** [comment]
Does this help?

<details>
  <summary>Spoiler: Click to reveal...</summary>

https://github.com/user-attachments/assets/db706488-506c-4f30-8582-ab3528247d0d
</details>

I'm afraid I cannot provide an unfiltered one as something else is spamming a generic warning that I'm trying to figure out now.

**Comment by KieronQuinn** [comment]
No, there's nothing in that log unfortunately

**Comment by YamatoRyou** [comment]
> It's opening 100% of the time for me on my Pixel 8 Pro. Can you get a log, see if it's crashing the times it's not opening?


The log is obtained by the following command:
```
adb shell
su
logcat | grep "com.kieronquinn.app.classicpowermenu"
```

Output (when the phone is locked but the screen is on, long press the power button to make the power menu flash once and then disappear. Only the content generated after the above action is completed is captured):
```
696  1465  4779 W Thanox-Core: AppLaunchStateManager, aboutToLaunch: Pkg{pkgName='com.kieronquinn.app.classicpowermenu', userId=0}
696  1465  4779 W Thanox-Core: mayEnableAppOnStartActivityIntent activityPackages: [com.kieronquinn.app.classicpowermenu]
696  1465  4779 W Thanox-Core: mayEnableAppOnStartActivityIntent pkg: Pkg{pkgName='com.kieronquinn.app.classicpowermenu', userId=0}
704  1465  4779 V SplashScreenExceptionList: SplashScreen checking exception for package com.kieronquinn.app.classicpowermenu (target sdk:35) -> false
711 23360 23393 V WindowManagerShell: Transition requested (#449): android.os.BinderProxy@eb563a TransitionRequestInfo { type = OPEN, triggerTask = TaskInfo{userId=0 taskId=672 displayId=0 isRunning=true baseIntent=Intent { flg=0x14830000 cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity } baseActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} topActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} origActivity=null realActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} numActivities=1 lastActiveTime=89764352 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{android.window.IWindowContainerToken$Stub$Proxy@35592eb} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=Rect(0, 123 - 0, 0) topActivityInfo=ActivityInfo{af55848 com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=false isVisibleRequested=false isSleeping=true locusId=null displayAreaFeatureId=1 isTopActivityTransparent=true appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false topActivityEligibleForLetterboxEducation= false isLetterboxDoubleTapEnabled= false topActivityEligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=1080 topActivityLetterboxHeight=2400 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false cameraCompatControlState=hidden}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 449 }
713  1465  4779 I ActivityTaskManager: START u0 {flg=0x14830000 cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity} with LAUNCH_SINGLE_INSTANCE from uid 0 (BAL_ALLOW_ALLOWLISTED_UID) result code=0
720  1465  2394 W Thanox-Core: reportOnActivityResumed: Intent { cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity }, user: 0
772  1465  2394 W Thanox-Core: reportOnActivityResumed: Intent { cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity }, user: 0
777  1465  5379 D CoreBackPreview: Window{b3c315 u0 com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity}: Setting back callback OnBackInvokedCallbackInfo{mCallback=android.window.IOnBackInvokedCallback$Stub$Proxy@aa181b8, mPriority=-1, mIsAnimationCallback=false}
788 23192 23192 D ControlsFavoritePersistenceWrapper: Reading data from file: /data/user/0/com.kieronquinn.app.classicpowermenu/files/controls_favorites.xml
814  1465  2384 I ActivityTaskManager: Displayed com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity for user 0: +108ms
815  1465  2384 V WindowManager: Sent Transition (#449) createdAt=698 via request=TransitionRequestInfo { type = OPEN, triggerTask = TaskInfo{userId=0 taskId=672 displayId=0 isRunning=true baseIntent=Intent { flg=0x14830000 cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity } baseActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} topActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} origActivity=null realActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} numActivities=1 lastActiveTime=89764352 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{RemoteToken{6c893e7 Task{7fef446 #672 type=standard A=10519:com.kieronquinn.app.classicpowermenu}}} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=Rect(0, 123 - 0, 0) topActivityInfo=ActivityInfo{20aa7a com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=false isVisibleRequested=false isSleeping=true locusId=null displayAreaFeatureId=1 isTopActivityTransparent=true appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false topActivityEligibleForLetterboxEducation= false isLetterboxDoubleTapEnabled= false topActivityEligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=1080 topActivityLetterboxHeight=2400 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false cameraCompatControlState=hidden}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 449 }
821  1465  2384 V WindowManager:         {WCT{RemoteToken{6c893e7 Task{7fef446 #672 type=standard A=10519:com.kieronquinn.app.classicpowermenu}}} m=OPEN f=TRANSLUCENT|NO_ANIMATION p=WCT{RemoteToken{8e77eef DefaultTaskDisplayArea@105330191}} leash=Surface(name=Task=672)/@0x3c06b01 sb=Rect(0, 0 - 1080, 2400) eb=Rect(0, 0 - 1080, 2400) d=0},
843 23360 23393 V WindowManagerShell: Transition requested (#450): android.os.BinderProxy@339f729 TransitionRequestInfo { type = CLOSE, triggerTask = TaskInfo{userId=0 taskId=672 displayId=0 isRunning=false baseIntent=Intent { flg=0x14830000 cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity } baseActivity=null topActivity=null origActivity=null realActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} numActivities=0 lastActiveTime=89764407 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{android.window.IWindowContainerToken$Stub$Proxy@bb2eaae} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=null topActivityInfo=null launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=true isVisibleRequested=true isSleeping=false locusId=null displayAreaFeatureId=1 isTopActivityTransparent=false appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false topActivityEligibleForLetterboxEducation= false isLetterboxDoubleTapEnabled= false topActivityEligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=-1 topActivityLetterboxHeight=-1 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false cameraCompatControlState=hidden}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 450 }
889  1465  2384 V WindowManager: Sent Transition (#450) createdAt=842 via request=TransitionRequestInfo { type = CLOSE, triggerTask = TaskInfo{userId=0 taskId=672 displayId=0 isRunning=false baseIntent=Intent { flg=0x14830000 cmp=com.kieronquinn.app.classicpowermenu/.ui.activities.PowerMenuActivity } baseActivity=null topActivity=null origActivity=null realActivity=ComponentInfo{com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity} numActivities=0 lastActiveTime=89764407 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{RemoteToken{6c893e7 Task{7fef446 #672 type=standard A=10519:com.kieronquinn.app.classicpowermenu}}} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=null topActivityInfo=null launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=true isVisibleRequested=true isSleeping=false locusId=null displayAreaFeatureId=1 isTopActivityTransparent=false appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false topActivityEligibleForLetterboxEducation= false isLetterboxDoubleTapEnabled= false topActivityEligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=-1 topActivityLetterboxHeight=-1 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false cameraCompatControlState=hidden}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 450 }
889  1465  2384 V WindowManager:         {WCT{RemoteToken{6c893e7 Task{7fef446 #672 type=standard A=10519:com.kieronquinn.app.classicpowermenu}}} m=CLOSE f=TRANSLUCENT p=WCT{RemoteToken{8e77eef DefaultTaskDisplayArea@105330191}} leash=Surface(name=Task=672)/@0x3c06b01 sb=Rect(0, 0 - 1080, 2400) eb=Rect(0, 0 - 1080, 2400) d=0},
907  1465  4779 D CoreBackPreview: Window{b3c315 u0 com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity}: Setting back callback null
912  1465  3109 W InputManager-JNI: Input channel object 'b3c315 com.kieronquinn.app.classicpowermenu/com.kieronquinn.app.classicpowermenu.ui.activities.PowerMenuActivity (client)' was disposed without first being removed with the input manager!
912  1465  3109 V WindowManager: Loading animations: layout params pkg=com.kieronquinn.app.classicpowermenu resId=0x1030308
949 14413 24261 I AppLocaleUtil: Can display preference - [com.kieronquinn.app.classicpowermenu] : isDisallowedPackage : false / isSignedWithPlatformKey : false / hasLauncherEntry : true / canDisplay : false / 1.1
```

**The content shows that other Xposed modules seem to be involved.**
BTW: each time the power button is long pressed, the menu flashes once. Repeat this operation at this time, and the menu will flash repeatedly.

**Comment by KieronQuinn** [comment]
Those logs show that the menu is being opened, and then closed again, so something else is definitely interfering here. No idea what. 

Nothing related to this stuff was changed between the two versions, other than _maybe_ increasing the target SDK version, but that shouldn't be causing this

**Comment by YamatoRyou** [comment]
> Those logs show that the menu is being opened, and then closed again, so something else is definitely interfering here. No idea what.
> 
> Nothing related to this stuff was changed between the two versions, other than _maybe_ increasing the target SDK version, but that shouldn't be causing this

I just tried the following:
1. Disable LSPosed in Magisk (which will also disable all Xposed modules);
2. Enable "Accessibility" permissions for apps and disable battery optimization;
3. Restart the phone;

Result:
Not only did it fail to solve the problem, but even pressing the power button on the desktop after unlocking it would cause the power menu to flicker.

After repeated tests, I found the following situations (all calculated from long pressing the power button):
a) The system's built-in power menu flickers;
b) The system's built-in power menu flickers, and then the app's power menu pops up and flickers;
c) The system's built-in power menu flickers, and then the app's power menu pops up and does not disappear (but this situation is rare).

When testing the above situations, the phone was unlocked and the desktop was displayed.
If the screen is locked but on, a and b will always appear.

**Comment by PowerRanger2130** [comment]
I'm having the same issue. The first time I open after restart it stays for a few seconds on lock screen, but after that it closes immediately.

**Comment by KieronQuinn** [comment]
I reverted the target SDK back down to Android 14, which is the only thing that I think may be causing this. Please try version 1.9

If it doesn't fix it, I'm all out of ideas.

**Comment by YamatoRyou** [comment]
> I reverted the target SDK back down to Android 14, which is the only thing that I think may be causing this. Please try version 1.9
> 
> If it doesn't fix it, I'm all out of ideas.

Unfortunately after updating to 1.9, the problem still exists. Here is a demonstration video.
https://github.com/user-attachments/assets/86e94fc7-4e4f-4eaf-8391-b2dde44ca0dd

**Comment by KieronQuinn** [comment]
I'm going to have to mark this as won't fix then, this revert was the only thing that might have fixed it. The other changes in 1.8 were all important library updates that I'm not prepared to revert.

**Comment by dariusz-goc** [comment]
This might be related to "Auto-switch default payment method to google pay when power menu is open" by @Philip2809. Can you revert that for testing?

**Comment by KieronQuinn** [comment]
> This might be related to "Auto-switch default payment method to google pay when power menu is open" by @Philip2809. Can you revert that for testing?

It wouldn't be built if they did. If you think it might be to blame, fork the repo, revert it and build the app. It won't be reverted for a mainline build.

**Comment by dariusz-goc** [comment]
> > This might be related to "Auto-switch default payment method to google pay when power menu is open" by @Philip2809. Can you revert that for testing?
> 
> It wouldn't be built if they did. If you think it might be to blame, fork the repo, revert it and build the app. It won't be reverted for a mainline build.

Not trying to offend anyone, it's just that Google made accessing the wallet require unlock quite recently, so changing the default service may be related. I'm happy to test a test build. Unfortunately at the moment I don't have the skills to revert and build myself, otherwise I would do that.

**Comment by JLJTGR** [comment]
I don't know if this means anything either way, but I don't use Google Pay at all.  The Play Store does not think I have Google Pay nor Google Wallet installed.  But this is a Pixel phone... so it is possible it is very much hiding until I try to activate something.  It doesn't show as a disabled/frozen app, though.  It does "appear" as a selection for contactless payments, which I do not use.

**Comment by Philip2809** [comment]
I am also facing this problem and I don't __think__ it is the Auto Switch, because I tried quickly by disabling the `Settings.Secure.putString` (which changes the default setting), removing all manifest permissions I added, etc. but it is still happening. Also, if you don't enable the auto-switch service then it should never call the `putString` function.  I will try to figure what causes this, but I might need some time as I am kinda busy the coming days

**Comment by dariusz-goc** [comment]
>I will try to figure what causes this, but I might need some time as I am kinda busy the coming days

Please do try. I woulda hope this app doesn't get abandoned, even thought the original developer seems to be losing interest in maintaining it. 

> Based on my logs it looks to me like it triggers a "navigate back" call, as it outputs the same when I open CPM unlocked and swipe to close and when I use it on the lockscreen.

Since version 1.8 swiping cards and expanding power options lags a bit for me. Maybe it's a hint for an infinite loop running? Maybe it's related to lockscreen disappearing. I downgraded to 1.7.2 for the time being. It fixes both lag and lockscreen visibility, which is the main reason for me why this app would be useful for me, to access home controls and payment options. If I have to unlock first then I might as well use any other shortcut to those things. I hope you will succeed in finding the issue 🤞

**Comment by KieronQuinn** [comment]
> even thought the original developer seems to be losing interest in maintaining it.

The app has been on maintenance mode about a year now, I only updated the libraries because it was required to build it on a modern Android Studio version. It's not getting feature updates from me, that's already been declared.

**Comment by KieronQuinn** [comment]
Loyalty cards was replaced by Smartspacer, controls & regular wallet is accessible from the lock screen still (it wasn't initially which is why this app was conceived). There's little reason for CPM anymore other than maintaining muscle memory.

**Comment by dariusz-goc** [comment]
> Loyalty cards was replaced by Smartspacer, controls & regular wallet is accessible from the lock screen still (it wasn't initially which is why this app was conceived). There's little reason for CPM anymore other than maintaining muscle memory.

That's true, but both of those use valuable space, either lockscreen shortcuts, which can be assigned different functions like qr code scanning or flashlight, or quick settings tiles which require extra gestures to get to. Using power menu, which in itself doesn't do much, was ingenious. Shame Google abandoned the idea, guess they were making space for yet another useless Gemini shortcut.

**Comment by KieronQuinn** [comment]
> Also accessing Wallet requires unlock now.

You can disable this with [GMS Flags](https://github.com/polodarb/GMS-Flags), set `GlobalActions__intent_to_wallet_home` to `false` in the `com.google.android.gms.tapandpay` package

**Comment by dariusz-goc** [comment]
> > Also accessing Wallet requires unlock now.
> 
> You can disable this with [GMS Flags](https://github.com/polodarb/GMS-Flags), set `GlobalActions__intent_to_wallet_home` to `false` in the `com.google.android.gms.tapandpay` package

Thanks, that's really helpful. However my guess is that Google will deprecate that flag soon enough.

**Comment by MillionsToOne** [comment]
Should be reopened.

**Comment by rom4ster** [comment]
I am facing this issue as well. Off the top of my head it looks like the app is trying to open and then closing itself

**Comment by rom4ster** [comment]
Ok So all this discussion and it is odd no one actually did any testing. I have done testing and have found the commit that introduces this. 

https://github.com/KieronQuinn/ClassicPowerMenu/commit/135342de93b965c3da0cf2244123eedcadfa525a


There are 67 files changed in that commit so if any of you folks looking at this would like to do some digging help appreciated.

**Comment by rom4ster** [comment]
Ok so we were completely off with the hunch on the wallet changes though its a good first place to check. 

It was most likely api level changes or something but I updated the manifest and the methodology for lock screen flags and I think its working on my phone where it wasnt before.

Check out my PR: 

https://github.com/KieronQuinn/ClassicPowerMenu/pull/156

**Comment by KieronQuinn** [comment]
Thanks, merged and built in 1.9.1. I _still_ cannot reproduce this, so you'll have to tell me if updating the dependencies & target back to 15 breaks it again. I've also got no idea why the fix even works, given all it's actually doing is moving where the window flags are set to the manifest - can only assume something in AppCompat or similar broke the previously working method on some devices.

**Comment by dariusz-goc** [comment]
1.9.1 fixed it for me on A14. Thanks guys!

**Comment by rom4ster** [comment]
> Thanks, merged and built in 1.9.1. I _still_ cannot reproduce this, so you'll have to tell me if updating the dependencies & target back to 15 breaks it again. I've also got no idea why the fix even works, given all it's actually doing is moving where the window flags are set to the manifest - can only assume something in AppCompat or similar broke the previously working method on some devices.

I have a (untested) theory, that the manifest entries needed to change but only the new methods would obey the new manifest. Thats a hunch though because honestly I have no idea why it works either.
